core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -489,8 +493,9 @@ class ReplicaManager(val config: KafkaConfig,
       val fetchPartitionStatus = logReadResults.map { case (topicAndPartition, result) =>
         (topicAndPartition, FetchPartitionStatus(result.info.fetchOffsetMetadata, fetchInfo.get(topicAndPartition).get))
       }
+
       val fetchMetadata = FetchMetadata(fetchMinBytes, fetchOnlyFromLeader, fetchOnlyCommitted, isFromFollower, fetchPartitionStatus)
-      val delayedFetch = new DelayedFetch(timeout, fetchMetadata, this, responseCallback)
+      val delayedFetch = new DelayedFetch(timeout, fetchMetadata, this, quota ,responseCallback)
```

Comment:
I did consider this issue. You are right that we could make the leader algorithm more responsive by altering the timeout (I didn't think of that - good idea). 
For now I'm inclined to raise a jira for this as a future enhancement. It shouldn't affect throttling significantly. The extra delay will even out over time.  
Also, this problem exists on the follower too so the optimisation is only of value for the leader side of the throttle. 
The main concern I have is actually the lack of smarts on the follower, particularly if throttled partitions enter the ISR, as the follower logic is very basic.